CONVENTIONS.md

```diff
@@ -0,0 +1,125 @@
+# Style conventions
+
+## Docstrings
+
+Docstrings shall be written in NumpyDoc format.
+
+Docstrings can either be just the summary text, or they must be complete.
+
+## Type annotations
+
+Type annotations will not be applied (for now).
+
+Note: I'm not very familiar with the latest and greatest type annotation
+options for libraries that make heavy use of numpy arrays like this one.
+
+## Linting
+
+Linting shall be performed with flake8, flake8-isort, flake8-black and pep8-naming.
+
+Automated formatting shall be performed with black and isort.
+
+Black is left at defaults, flake8 and isort are configured to adhere to black.
+
+## Testing
+
+All functions need to be covered by unit tests.
+
+
+# Functional API conventions
+
+This API is for internal use and for power-users that want to
+vectorize operations on large sets of things.
+
+Performance is prioritized over extensive input validation.
+
+The source for this API resides in the `pylinalg.func` subpackage and is organized
+by type.
+
+## Function naming
+
+The functional API has rather verbose names, but it makes things
+explicit.
+
+## Function signatures
+
+We strive to align closely with the conventions of [numpy](https://numpy.org/doc/stable/reference/ufuncs.html#ufuncs-kwargs)
+in order to be least-surprising for pylinalg users.
+
+All functions shall adhere as much possible to the following signature:
+
+`func(x1, x2, ..., /, *, out=None, dtype=None, **kwargs)`
+
+* All input arguments are positional-only.
+* There can be as many input arguments as needed for the function signature to make sense.
+* Input arguments are always [`array_like`](https://numpy.org/doc/stable/glossary.html?highlight=array_like#term-array_like):
```

Comment:
I like the general idea behind this, but I fear it might be too restrictive. I can see scenarios where you want something like a `method="linear"` parameter or where you need a `eps=1e-6`. 

Of course, you can require that those are arrays and apply broadcasting to them to match the other inputs, but I am not sure if this adds value in comparison to making faulty inputs harder to catch.

Reply:
IIUC Args like these would be keyword arguments. A bullet further down says "There may be additionaleyword-only arguments as required by the specifics of individual functions."

Reply:
Almar is correct. I tried to clarify this.